README.md
# mlr3learners.kernlab

<!-- badges: start -->

[![R CMD Check via {tic}](https://github.com/mlr3learners/mlr3learners.kernlab/workflows/R%20CMD%20Check%20via%20{tic}/badge.svg?branch=master)](https://github.com/mlr3learners/mlr3learners.kernlab/actions)
![Parameter Check](https://github.com/mlr3learners/mlr3learners.kernlab/workflows/Parameter%20Check/badge.svg?branch=master)
[![codecov](https://codecov.io/gh/mlr3learners/mlr3learners.kernlab/branch/master/graph/badge.svg)](https://codecov.io/gh/mlr3learners/mlr3learners.kernlab)
[![StackOverflow](https://img.shields.io/badge/stackoverflow-mlr3-orange.svg)](https://stackoverflow.com/questions/tagged/mlr3)
<!-- badges: end -->

Adds `ksvm()` from the {kernlab} package to {mlr3}.

Install the latest release of the package via 

```r
install.packages("mlr3learners.kernlab")
```

by following the instructions in the [mlr3learners.drat README](https://github.com/mlr3learners/mlr3learners.drat).

Alternatively, you can install the latest version of {mlr3learners.kernlab} from Github with:

```r
remotes::install_github("mlr3learners/mlr3learners.kernlab")
```

## Supported kernels

- `rbfdot`
- `polydot`
- `vanilladot`
- `laplacedot`
- `besseldot`
- `anovadot`

## Currently unsupported kernels

- `tanhdot`
- `splinedot`
- `stringdot`


NEWS.md
# mlr3learners.kernlab 0.2.0.9000

- Internal changes only.


# mlr3learners.kernlab 0.2.0

- Add Paramtest
- Rename package from "mlr3learners.ksvm" to "mlr3learners.kernlab"
- Add lintr

# mlr3learners.kernlab 0.1.1.9000

- Internal changes only

# mlr3learners.kernlab 0.1.1

- Update documentation, methods and DESCRIPTION
- Add man for help method
- gh actions workflow
- Remove travis
- Remove appveyor
- Update zzz.R
- Add to drat

# mlr3learners.kernlab 0.1.0.9000

- Internal changes only.

# mlr3learner.kernlab 0.1.0

- Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(mlr3learners.kernlab)

test_check("mlr3learners.kernlab")


tests/testthat/test_ksvm_kpar.R
context("ksvm.kpar")

test_kpar = function(kernel, kpar) {
  learner = do.call(lrn, args = append(list("regr.ksvm", kernel = kernel), kpar))
  expect_learner(learner)
  result = run_autotest(learner)
  expect_true(result, info = result$error)

  learner = do.call(lrn, args = append(list("classif.ksvm", kernel = kernel), kpar))
  expect_learner(learner)
  result = run_autotest(learner)
  expect_true(result, info = result$error)
}

# working:
test_kpar("rbfdot", list(sigma = 5))
test_kpar("polydot", list(degree = 5, scale = 2, offset = 10))
test_kpar("vanilladot", list())
test_kpar("laplacedot", list(sigma = 5))
test_kpar("anovadot", list(sigma = 5, degree = 3))
test_kpar("besseldot", list(sigma = 5, degree = 2, order = 2))

# broken:
# test_kpar("tanhdot", list(scale = 1, offset = 2))
# test_kpar("splinedot", list())

tests/testthat/test_regr_ksvm.R
context("regr.ksvm")

test_that("autotest", {
  learner = LearnerRegrKSVM$new()
  expect_learner(learner)
  result = run_autotest(learner)
  expect_true(result, info = result$error)
})


R/zzz.R
#' @import data.table
#' @import paradox
#' @import mlr3misc
#' @importFrom R6 R6Class
#' @importFrom mlr3 mlr_learners LearnerClassif LearnerRegr
"_PACKAGE"

# nocov start
register_mlr3 = function(libname, pkgname) {
  # get mlr_learners dictionary from the mlr3 namespace
  x = utils::getFromNamespace("mlr_learners", ns = "mlr3")

  # add the learner to the dictionary
  x$add("classif.ksvm", LearnerClassifKSVM)
  x$add("regr.ksvm", LearnerRegrKSVM)
}

.onLoad = function(libname, pkgname) { # nolint
  register_mlr3()
  setHook(packageEvent("mlr3", "onLoad"), function(...) register_mlr3(),
    action = "append")
}

.onUnload = function(libpath) { # nolint
  event = packageEvent("mlr3", "onLoad")
  hooks = getHook(event)
  pkgname = vapply(hooks, function(x) environment(x)$pkgname, NA_character_)
  setHook(event, hooks[pkgname != "mlr3learners.kernlab"],
    action = "replace")
}
# nocov end


R/LearnerClassifKSVM.R
#' @title Classification Kernlab Support Vector Machine
#'
#' @name mlr_learners_classif.ksvm
#'
#' @description
#' Classification support vector machine.
#' Calls  [kernlab::ksvm()] from package \CRANpkg{kernlab}.
#'
#' @templateVar id classif.ksvm
#' @template section_dictionary_learner
#'
#' @references
#' \cite{mlr3learners.kernlab}{karatzoglou_2004}
#'
#' @export
#' @template seealso_learner
#' @template example
LearnerClassifKSVM = R6Class("LearnerClassifKSVM",
  inherit = LearnerClassif,
  public = list(

    #' @description
    #' Creates a new instance of this [R6][R6::R6Class] class.
    initialize = function() {
      ps = ParamSet$new(list(
        ParamLgl$new(id = "scaled", default = TRUE, tags = c("train")),
        ParamFct$new(
          id = "type", default = "C-svc",
          levels = c("C-svc", "nu-svc", "C-bsvc", "spoc-svc", "kbb-svc"),
          tags = c("train")),
        ParamFct$new(
          id = "kernel", default = "rbfdot",
          levels = c(
            "rbfdot", "polydot", "vanilladot",
            "laplacedot", "besseldot", "anovadot"),
          tags = c("train")),
        ParamDbl$new(id = "C", default = 1, tags = c("train")),
        ParamDbl$new(id = "nu", default = 0.2, lower = 0, tags = c("train")),
        ParamInt$new(id = "cache", default = 40, lower = 1L, tags = c("train")),
        ParamDbl$new(id = "tol", default = 0.001, lower = 0, tags = c("train")),
        ParamLgl$new(id = "shrinking", default = TRUE, tags = c("train")),
        ParamDbl$new(id = "sigma", default = NO_DEF, lower = 0, tags = "train"),
        ParamInt$new(
          id = "degree", default = NO_DEF, lower = 1L,
          tags = "train"),
        ParamDbl$new(id = "scale", default = NO_DEF, lower = 0, tags = "train"),
        ParamInt$new(id = "order", default = NO_DEF, tags = "train"),
        ParamDbl$new(id = "offset", default = NO_DEF, tags = "train")
      ))

      ps$add_dep(
        "C", "type",
        CondAnyOf$new(c("C-svc", "C-bsvc", "spoc-svc", "kbb-svc")))
      ps$add_dep("nu", "type", CondAnyOf$new(c("nu-svc")))
      ps$add_dep(
        "sigma", "kernel",
        CondAnyOf$new(c("rbfdot", "laplacedot", "besseldot", "anovadot")))
      ps$add_dep(
        "degree", "kernel",
        CondAnyOf$new(c("polydot", "besseldot", "anovadot")))
      ps$add_dep("scale", "kernel", CondAnyOf$new(c("polydot")))
      ps$add_dep("order", "kernel", CondAnyOf$new(c("besseldot")))
      ps$add_dep("offset", "kernel", CondAnyOf$new(c("polydot")))

      super$initialize(
        id = "classif.ksvm",
        packages = "kernlab",
        feature_types = c(
          "logical", "integer", "numeric",
          "character", "factor", "ordered"),
        predict_types = c("response", "prob"),
        param_set = ps,
        properties = c("weights", "twoclass", "multiclass"),
        man = "mlr3learners.kernlab::mlr_learners_classif.ksvm"
      )
    }),

  private = list(
    .train = function(task) {
      pars = self$param_set$get_values(tags = "train")
      kpar = intersect(
        c("sigma", "degree", "scale", "order", "offset"),
        names(pars))

      if ("weights" %in% task$properties) {
        pars$class.weights = task$weights$weight
      }

      if (length(kpar) > 0) {
        pars$kpar = pars[kpar]
        pars[kpar] = NULL
      }

      f = task$formula()
      data = task$data()

      invoke(kernlab::ksvm,
        x = f, data = data,
        prob.model = self$predict_type == "prob", .args = pars)
    },

    .predict = function(task) {
      newdata = task$data(cols = task$feature_names)

      predict_type = ifelse(self$predict_type == "prob",
        "probabilities", "response")
      p = invoke(kernlab::predict, self$model,
        newdata = newdata,
        type = predict_type)

      if (self$predict_type == "response") {
        PredictionClassif$new(task = task, response = p)
      } else {
        PredictionClassif$new(task = task, prob = p)
      }
    }
  )
)


R/LearnerRegrKSVM.R
#' @title Regression Kernlab Support Vector Machine
#'
#' @name mlr_learners_regr.ksvm
#'
#' @description
#' Regression support vector machine.
#' Calls  [kernlab::ksvm()] from package \CRANpkg{kernlab}.
#'
#' @templateVar id regr.ksvm
#' @template section_dictionary_learner
#'
#' @references
#' \cite{mlr3learners.kernlab}{karatzoglou_2004}
#'
#' @export
#' @template seealso_learner
#' @template example
LearnerRegrKSVM = R6Class("LearnerRegrKSVM",
  inherit = LearnerRegr,
  public = list(

    #' @description
    #' Creates a new instance of this [R6][R6::R6Class] class.
    initialize = function() {
      ps = ParamSet$new(list(
        ParamLgl$new(id = "scaled", default = TRUE, tags = "train"),
        ParamFct$new(
          id = "type", default = "eps-svr",
          levels = c("eps-svr", "nu-svr", "eps-bsvr"), tags = "train"),
        ParamFct$new(
          id = "kernel", default = "rbfdot",
          levels = c(
            "rbfdot", "polydot", "vanilladot",
            "laplacedot", "besseldot", "anovadot"),
          tags = "train"),
        ParamDbl$new(id = "C", default = 1, tags = "train"),
        ParamDbl$new(id = "nu", default = 0.2, lower = 0, tags = "train"),
        ParamDbl$new(id = "epsilon", default = 0.1, tags = "train"),
        ParamInt$new(id = "cache", default = 40, lower = 1L, tags = "train"),
        ParamDbl$new(id = "tol", default = 0.001, lower = 0, tags = "train"),
        ParamLgl$new(id = "shrinking", default = TRUE, tags = "train"),
        ParamDbl$new(id = "sigma", default = NO_DEF, lower = 0, tags = "train"),
        ParamInt$new(
          id = "degree", default = NO_DEF, lower = 1L,
          tags = "train"),
        ParamDbl$new(id = "scale", default = NO_DEF, lower = 0, tags = "train"),
        ParamInt$new(id = "order", default = NO_DEF, tags = "train"),
        ParamDbl$new(id = "offset", default = NO_DEF, tags = "train")
      ))

      ps$add_dep("C", "type", CondAnyOf$new(c("eps-svr", "eps-bsvr")))
      ps$add_dep("nu", "type", CondAnyOf$new(c("nu-svr")))
      ps$add_dep(
        "epsilon", "type",
        CondAnyOf$new(c("eps-svr", "nu-svr", "eps-bsvr")))

      ps$add_dep(
        "sigma", "kernel",
        CondAnyOf$new(c("rbfdot", "laplacedot", "besseldot", "anovadot")))
      ps$add_dep(
        "degree", "kernel",
        CondAnyOf$new(c("polydot", "besseldot", "anovadot")))
      ps$add_dep("scale", "kernel", CondAnyOf$new(c("polydot")))
      ps$add_dep("order", "kernel", CondAnyOf$new(c("besseldot")))
      ps$add_dep("offset", "kernel", CondAnyOf$new(c("polydot")))

      super$initialize(
        id = "regr.ksvm",
        packages = "kernlab",
        feature_types = c(
          "logical", "integer", "numeric",
          "character", "factor", "ordered"),
        predict_types = "response",
        param_set = ps,
        properties = "weights",
        man = "mlr3learners.kernlab::mlr_learners_regr.ksvm"
      )
    }),

  private = list(
    .train = function(task) {
      pars = self$param_set$get_values(tags = "train")
      kpar = intersect(
        c("sigma", "degree", "scale", "order", "offset"),
        names(pars))

      if ("weights" %in% task$properties) {
        pars$class.weights = task$weights$weight
      }

      if (length(kpar) > 0) {
        pars$kpar = pars[kpar]
        pars[kpar] = NULL
      }

      f = task$formula()
      data = task$data()

      invoke(kernlab::ksvm, x = f, data = data, .args = pars)
    },

    .predict = function(task) {
      newdata = task$data(cols = task$feature_names)

      p = invoke(kernlab::predict, self$model,
        newdata = newdata,
        type = "response")
      PredictionRegr$new(task = task, response = p)
    }
  )
)
